File: org-hr-system/server/src/department/department.controller.ts
import {
  Controller,
  Get,
  Post,
  Put,
  Delete,
  Body,
  Param,
  Query,
  ParseIntPipe,
} from '@nestjs/common';
import { DepartmentService } from './department.service';
import { CreateDepartmentDto } from './dto/create-department.dto';
import { MoveDepartmentDto } from './dto/move-department.dto';

@Controller('api/departments')
export class DepartmentController {
  constructor(private readonly departmentService: DepartmentService) {}

  /**
   * GET /api/departments/tree
   * 查询完整树或按 parentId 懒加载子节点
   */
  @Get('tree')
  async getTree(@Query('parentId') parentId?: string) {
    if (parentId) {
      return this.departmentService.findChildren(Number(parentId));
    }
    return this.departmentService.findAll();
  }

  /**
   * POST /api/departments
   * 创建部门
   */
  @Post()
  async create(@Body() dto: CreateDepartmentDto) {
    return this.departmentService.create(dto);
  }

  /**
   * PUT /api/departments/:id
   * 更新部门名称
   */
  @Put(':id')
  async update(
    @Param('id', ParseIntPipe) id: number,
    @Body('name') name: string,
  ) {
    return this.departmentService.update(id, name);
  }

  /**
   * DELETE /api/departments/:id
   * 删除部门
   */
  @Delete(':id')
  async delete(@Param('id', ParseIntPipe) id: number) {
    await this.departmentService.delete(id);
    return { message: '删除成功' };
  }

  /**
   * PUT /api/departments/:id/move
   * 移动部门
   */
  @Put(':id/move')
  async move(
    @Param('id', ParseIntPipe) id: number,
    @Body() dto: MoveDepartmentDto,
  ) {
    return this.departmentService.move(id, dto.newParentId);
  }
}

File: src/main/java/com/example/orgmgmt/controller/DepartmentController.java
package com.example.orgmgmt.controller;

import com.example.orgmgmt.common.ApiResponse;
import com.example.orgmgmt.dto.DepartmentCreateRequest;
import com.example.orgmgmt.dto.DepartmentDTO;
import com.example.orgmgmt.dto.DepartmentMoveRequest;
import com.example.orgmgmt.dto.DepartmentUpdateRequest;
import com.example.orgmgmt.entity.Department;
import com.example.orgmgmt.service.DepartmentService;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;

import java.util.List;

/**
 * 部门管理控制器
 */
@RestController
@RequestMapping("/api/departments")
@RequiredArgsConstructor
public class DepartmentController {

    private final DepartmentService departmentService;

    /**
     * 获取部门树
     */
    @PreAuthorize("hasAnyRole('ADMIN', 'HR', 'DEPT_MANAGER')")
    @GetMapping("/tree")
    public ApiResponse<List<DepartmentDTO>> getTree() {
        List<DepartmentDTO> tree = departmentService.getTree();
        return ApiResponse.success(tree);
    }

    /**
     * 获取指定部门的子部门列表（懒加载）
     */
    @PreAuthorize("hasAnyRole('ADMIN', 'HR', 'DEPT_MANAGER')")
    @GetMapping("/{id}/children")
    public ApiResponse<List<DepartmentDTO>> getChildren(@PathVariable Long id) {
        List<DepartmentDTO> children = departmentService.getChildren(id);
        return ApiResponse.success(children);
    }

    /**
     * 创建部门
     */
    @PreAuthorize("hasRole('ADMIN')")
    @PostMapping
    public ApiResponse<Department> create(@Valid @RequestBody DepartmentCreateRequest request) {
        Department dept = departmentService.create(request);
        return ApiResponse.success(dept);
    }

    /**
     * 更新部门
     */
    @PreAuthorize("hasRole('ADMIN')")
    @PutMapping("/{id}")
    public ApiResponse<Department> update(@PathVariable Long id, @Valid @RequestBody DepartmentUpdateRequest request) {
        Department dept = departmentService.update(id, request);
        return ApiResponse.success(dept);
    }

    /**
     * 删除部门（逻辑删除）
     */
    @PreAuthorize("hasRole('ADMIN')")
    @DeleteMapping("/{id}")
    public ApiResponse<String> delete(@PathVariable Long id) {
        departmentService.delete(id);
        return ApiResponse.success("删除成功");
    }

    /**
     * 移动部门
     */
    @PreAuthorize("hasRole('ADMIN')")
    @PutMapping("/{id}/move")
    public ApiResponse<String> move(@PathVariable Long id, @Valid @RequestBody DepartmentMoveRequest request) {
        departmentService.move(id, request);
        return ApiResponse.success("移动成功");
    }
}
